### PR TITLE
SVM Attack

### PR DESCRIPTION
**Pull request by ebubae**
# Description

Added SVM attack for linear, polynomial and RBF kernels according to [Biggio et. al. Poisoning Attacks against Support Vector Machines](https://arxiv.org/pdf/1206.6389.pdf)

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Tested attack on Scikitlearn's LinearSVC and SVC with linear, polynomial, and RBF kernels

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/svm_attack.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+
+from art.attacks.attack import Attack
+from art.classifiers.scikitlearn import ScikitlearnSVC
+
+from sklearn.preprocessing import normalize
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: To minimise dependencies, could you please move this import to line 164?

File: art/attacks/svm_attack.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+from sklearn.svm import LinearSVC, SVC
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: To minimise dependencies, could you please move these imports after the docstrings of the methods that require them?

File: art/attacks/svm_attack.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        super(SVMAttack, self).set_params(**kwargs)
+        if self.step <= 0:
+            raise ValueError("Step size must be strictly positive")
+        if self.eps <= 0:
+            raise ValueError("Value of eps must be strictly positive")
+        if self.max_iters <= 1:
+            raise ValueError("Value of max_iters must be strictly positive")
+
+
+def generate_attack_point(classifier, x_train, y_train, x_val, y_val, x_attack, y_attack, step, eps, max_iters):
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: Could you make this an instance method belonging to the class?
If yes, remove the redundant arguments and use `self`.
It looks like this function is called only from within this class. If you don't intend users to call this function, you could add a underscore `_` at the beginning of the method name.

File: art/attacks/svm_attack.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+    poisoned_input = np.vstack([x_train, attack_point])
+    poisoned_labels = np.append(y_t, y_a)
+    poisoned_model.fit(poisoned_input, poisoned_labels)
+    poisoned_model = ScikitlearnSVC(model=poisoned_model, clip_values=classifier.clip_values,
+                                    defences=classifier.defences, preprocessing=classifier.preprocessing)
+    return attack_point, poisoned_model
+
+
+def predict_sign(model, input):
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: Could you make this an instance method belonging to the class?
If yes, remove the redundant arguments and use `self`.
It looks like this function is called only from within this class. If you don't intend users to call this function, you could add a underscore `_` at the beginning of the method name.

File: art/attacks/svm_attack.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    preds = model.predict(input)
+    one = 1
+    zero = 0
+    signs = np.zeros(preds.shape[0], )
+    signs[preds == one] = 1
+    signs[preds == zero] = -1
+    return signs
+
+
+def attack_gradient(model, attack_point, valid_data, valid_labels):
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: Could you make this an instance method belonging to the class?
If yes, remove the redundant arguments and use `self`.
It looks like this function is called only from within this class. If you don't intend users to call this function, you could add a underscore `_` at the beginning of the method name.

File: art/attacks/svm_attack.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        self.y_val = y_val
+        self.max_iters = max_iters
+        self.set_params(**kwargs)
+
+    def generate(self, x, **kwargs):
+        """
+        Iteratively finds optimal attack points starting at values at x
+
+        :param x: An array with the points that initialize attack points.
+        :type x: `np.ndarray`
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: I think we should add `y` to the docstring to show that it can be provided by the user.

File: art/attacks/svm_attack.py (hunk 3 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: I thought it could be used as a convenience function and is called explicitly from the notebook.

File: art/attacks/svm_attack.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        else:
+            y_attack = self.classifier.predict(x)
+
+        num_poison = len(x)
+        if num_poison == 0:
+            raise ValueError("Must input at least one poison point")
+        num_features = len(x[0])
+        train_data = np.copy(self.x_train)
+        train_labels = np.copy(self.y_train)
+        classifier = self.classifier
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]: Is this step necessary? I think it would refer to the same instance of classifier.

Comment (ebubae) [reply art/attacks/svm_attack.py]: It wouldn't. New instances of the classifier are being created by `generate_attack_point`

File: docs/modules/attacks.rst [hunk 8]
```diff
@@ -74,6 +74,12 @@ Spatial Transformations Attack
    :members:
    :special-members:
 
+SVM Attack
+------------------------------
```
Comment (beat-buesser) [inline docs/modules/attacks.rst]: Remove dashes
```suggestion
----------
```

File: art/classifiers/scikitlearn.py [hunk 9]
```diff
@@ -860,11 +860,14 @@ def __init__(self, model, clip_values=None, defences=None, preprocessing=(0, 1))
...
         if not isinstance(model, SVC) and not isinstance(model, LinearSVC):
-            raise TypeError('Model must be of type sklearn.svm.SVC or sklearn.svm.LinearSVC')
+            raise TypeError('Model must be of type sklearn.svm.SVC or sklearn.svm.LinearSVC. Found type {}'
+                            .format(type(model)))
 
         super(ScikitlearnSVC, self).__init__(model=model, clip_values=clip_values, defences=defences,
                                              preprocessing=preprocessing)
         self._model = model
+        self.kernel_func = self._kernel_func()
+        self.kernel_grad = self._kernel_gradient_func
```
Comment (beat-buesser) [inline art/classifiers/scikitlearn.py]: What is the reason for the changes in this and the previous line? Would it be possible to just use the function names on the right hand side?

File: art/classifiers/scikitlearn.py [hunk 10]
```diff
@@ -997,6 +1018,59 @@ def loss_gradient(self, x, y, **kwargs):
...
+
+    def q_submatrix(self, rows, cols):
+        """
+        Returns the q submatrix of this SVM indexed by the arrays at rows and columns
+        :param rows: the row vectors
+        :type rows: `np.ndarray`
+        :param cols: the column vectors
+        :type cols: `np.ndarray`
+        :return: a submatrix of Q
+        :rtype: np.ndarray
```
Comment (beat-buesser) [inline art/classifiers/scikitlearn.py]:
```suggestion
        :rtype: `np.ndarray`
```

File: art/classifiers/scikitlearn.py [hunk 11]
```diff
@@ -997,6 +1018,59 @@ def loss_gradient(self, x, y, **kwargs):
 
         return gradients
 
+    def _kernel_func(self):
+        """
+        Return the function for the kernel of this SVM
+        :param kernel: A string preloaded function or callable Python function
+        :type kernel: str or callable
```
Comment (beat-buesser) [inline art/classifiers/scikitlearn.py]:
```suggestion
        :type kernel: `str` or callable
```

File: art/classifiers/scikitlearn.py [hunk 12]
```diff
@@ -885,26 +888,44 @@ def class_gradient(self, x, label=None, **kwargs):
...
         else:
             raise NotImplementedError(
                 'Loss gradients for kernel \'{}\' are not implemented.'.format(self._model.kernel))
         return grad
 
+    def _get_kernel_gradient_sv(self, i_sv, x_sample):
+        """
+                Applies the kernel gradient to all of a model's support vectors
+                :param i_sv: A support vector index
+                :type i_sv: int
```
Comment (beat-buesser) [inline art/classifiers/scikitlearn.py]:
```suggestion
                :type i_sv: `int`
```

File: art/attacks/svm_attack.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    :param y_val: validation labels
+    :type y_val: `np.ndarray`
+    :param x_attack: the initial attack point
+    :type x_attack: `np.ndarray`
+    :param y_attack: the initial attack label
+    :type y_attack: `np.ndarray`
+    :param step: the step size
+    :type step: `float`
+    :param eps: the min difference in loss to aim for
+    :type eps: float
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
    :type eps: `float`
```

File: art/attacks/svm_attack.py [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    :param max_iters: the maximum number of iterations the attack should run
+    :type max_iters: int
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
    :type max_iters: `int`
```

File: art/attacks/svm_attack.py [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+            raise ValueError("Value of eps must be strictly positive")
+        if self.max_iters <= 1:
+            raise ValueError("Value of max_iters must be strictly positive")
+
+
+def generate_attack_point(classifier, x_train, y_train, x_val, y_val, x_attack, y_attack, step, eps, max_iters):
+    """
+    Generate a single poison attack the model, using x_val and y_val as validation points.
+    The attack begins at the point init_attack. The attack class will be the opposite of the model's
+    classification for init_attack.
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
    classification for `init_attack`.
```

File: art/attacks/svm_attack.py [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+            raise ValueError("Step size must be strictly positive")
+        if self.eps <= 0:
+            raise ValueError("Value of eps must be strictly positive")
+        if self.max_iters <= 1:
+            raise ValueError("Value of max_iters must be strictly positive")
+
+
+def generate_attack_point(classifier, x_train, y_train, x_val, y_val, x_attack, y_attack, step, eps, max_iters):
+    """
+    Generate a single poison attack the model, using x_val and y_val as validation points.
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
    Generate a single poison attack the model, using `x_val` and `y_val` as validation points.
```

File: art/attacks/svm_attack.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+    attack_params = ['classifier', 'step', 'eps', 'x_train', 'y_train', 'x_val', 'y_val']
+
+    def __init__(self, classifier, step, eps, x_train, y_train, x_val=None, y_val=None, max_iters=100, **kwargs):
+        """
+        Initialize an SVM attack
+
+        :param classifier: A trained ScikitlearnSVC classifier
+        :type classifier: `art.classifiers.scikitlearn.ScikitlearnSVC`
+        :param step: The step size of the classifier
+        :type step: float
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
        :type step: `float`
```

File: art/attacks/svm_attack.py [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        :param x_train: The training data used for classification
+        :type x_train: `np.ndarray`
+        :param y_train: The training labels used for classification
+        :type y_train: `np.ndarray`
+        :param x_val: The validation data used to test the attack
+        :type x_val: `np.ndarray`
+        :param y_val: The validation labels used to test the attack
+        :type y_val: `np.ndarray`
+        :param max_iters: The maximum number of iterations for the attack
+        :type max_iters: int
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
        :type max_iters: `int`
```

File: art/attacks/svm_attack.py [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+    def __init__(self, classifier, step, eps, x_train, y_train, x_val=None, y_val=None, max_iters=100, **kwargs):
+        """
+        Initialize an SVM attack
+
+        :param classifier: A trained ScikitlearnSVC classifier
+        :type classifier: `art.classifiers.scikitlearn.ScikitlearnSVC`
+        :param step: The step size of the classifier
+        :type step: float
+        :param eps: The minimum difference in loss before convergence of the classifier
+        :type eps: float
```
Comment (beat-buesser) [inline art/attacks/svm_attack.py]:
```suggestion
        :type eps: `float`
```

File: art/classifiers/scikitlearn.py (hunk 9 above)
Comment (ebubae) [reply art/classifiers/scikitlearn.py]: For `kernel_grad` yes it's essentially just a renaming, but `_kernel_func()` is a function that returns a function (the appropriate kernel). I'll still add the underscores.

File: art/attacks/svm_attack.py (hunk 4 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: `generate_attack_point` (a static function) also depends on this, so it too must be static

File: art/attacks/svm_attack.py (hunk 5 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: `generate_attack_point` (a static function) also depends on this, so it too must be static

File: art/attacks/svm_attack.py (hunk 3 above)
Comment (beat-buesser) [reply art/attacks/svm_attack.py]: I see. To follow the pattern of the other attacks in ART, could we make it an instance method and call it in the notebook as something like ```
attack = SVMAttack()
attack_point, poisoned = attack.generate_attack_point()
```

File: art/attacks/svm_attack.py (hunk 6 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/classifiers/scikitlearn.py (hunk 11 above)
Comment (ebubae) [reply art/classifiers/scikitlearn.py]: addressed in commit

File: art/classifiers/scikitlearn.py (hunk 12 above)
Comment (ebubae) [reply art/classifiers/scikitlearn.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 13 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 15 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 16 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 17 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in comit

File: art/attacks/svm_attack.py (hunk 18 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/classifiers/scikitlearn.py (hunk 10 above)
Comment (ebubae) [reply art/classifiers/scikitlearn.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 14 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in commit

File: art/attacks/svm_attack.py (hunk 19 above)
Comment (ebubae) [reply art/attacks/svm_attack.py]: addressed in com mit

**Comment by beat-buesser** [comment]
Hi @ebubae Thank you very much for your contributing the SVM poisoning attack to ART!
